From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chook <joshuawereszczuk@gmail.com>
Date: Wed, 8 Dec 2021 13:42:08 +1000
Subject: [PATCH] Adding Configurable Gravity Block Duplication


diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index cd918cec00d8202252af0d20b1a8891371c538e3..78db3a779005cac5e565fda34058df21b13877bc 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -480,6 +480,17 @@ public class PaperConfig {
         allowBlockPermanentBreakingExploits = getBoolean("settings.unsupported-settings.allow-permanent-block-break-exploits", allowBlockPermanentBreakingExploits);
     }
 
+    public static boolean allowGravityBlockDuplicationExploits = true;
+    private static void allowGravityBlockDuplicationExploits() {
+        if (config.contains("allow-gravity-block-duplication-exploits")) {
+            allowGravityBlockDuplicationExploits = config.getBoolean("allow-gravity-block-duplication-exploits", true);
+            config.set("allow-gravity-block-duplication-exploits", null);
+        }
+
+        config.set("settings.sandpaper-settings.allow-gravity-block-duplication-exploits-readme", "This setting controls if players should be able use gravity block duplication exploits through the use of end portals.");
+        allowGravityBlockDuplicationExploits = getBoolean("settings.sandpaper-settings.allow-gravity-block-duplication-exploits", allowGravityBlockDuplicationExploits);
+    }
+
     public static boolean consoleHasAllPermissions = false;
     private static void consoleHasAllPermissions() {
         consoleHasAllPermissions = getBoolean("settings.console-has-all-permissions", consoleHasAllPermissions);
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 22c1699a94c9ee1d0baa0eea3f30718cb433d52f..ed6927ae7a88a32071595691ea0329fd8c861774 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2746,7 +2746,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
             }
 
             this.processPortalCooldown();
-            //this.tickEndPortal(); Paper - make end portalling safe
+            // SandPaper start - add configurable gravity block dupe
+            if (!com.destroystokyo.paper.PaperConfig.allowGravityBlockDuplicationExploits) {
+                this.tickEndPortal(); // Paper - make end portalling safe
+            }
+            // SandPaper end - add configurable gravity block dupe
         }
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index 0c94b4cb6ee0c3dffe0b67a2291d0160ae0ef96f..c4927775f388484ec23d62e82833a3f6f1df8030 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -110,11 +110,16 @@ public class FallingBlockEntity extends Entity {
 
     @Override
     public void tick() {
-        // Paper start - fix sand duping
-        if (this.isRemoved()) {
-            return;
+        // SandPaper start - add configurable gravity block dupe
+        if (!com.destroystokyo.paper.PaperConfig.allowGravityBlockDuplicationExploits) {
+            // Paper start - fix sand duping
+            if (this.isRemoved()) {
+                return;
+            }
+            // Paper end - fix sand duping
         }
-        // Paper end - fix sand duping
+        // SandPaper end - add configurable gravity block dupe
+
         if (this.blockState.isAir()) {
             this.discard();
         } else if (this.level.isClientSide && this.removeAtMillis > 0L) {
@@ -152,11 +157,15 @@ public class FallingBlockEntity extends Entity {
 
             this.move(MoverType.SELF, this.getDeltaMovement());
 
-            // Paper start - fix sand duping
-            if (this.isRemoved()) {
-                return;
+            // SandPaper start - add configurable gravity block dupe
+            if (!com.destroystokyo.paper.PaperConfig.allowGravityBlockDuplicationExploits) {
+                // Paper start - fix sand duping
+                if (this.isRemoved()) {
+                    return;
+                }
+                // Paper end - fix sand duping
             }
-            // Paper end - fix sand duping
+            // SandPaper end - add configurable gravity block dupe
 
             // Paper start - Configurable EntityFallingBlock height nerf
             if (this.level.paperConfig.fallingBlockHeightNerf != 0 && this.getY() > this.level.paperConfig.fallingBlockHeightNerf) {
diff --git a/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java b/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
index 51456de352a79ebe68f71bdfb181773ced368c31..87b791ed341dfcf6dcf8bce971977f1097e510c5 100644
--- a/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
@@ -53,10 +53,25 @@ public class EndPortalBlock extends BaseEntityBlock {
                 // return; // CraftBukkit - always fire event in case plugins wish to change it
             }
 
-            // Paper start - move all of this logic into portal tick
-            //entity.portalWorld = ((ServerLevel)world);
-            //entity.portalBlock = pos.immutable();
-            // Paper end - move all of this logic into portal tick
+            // SandPaper start - add configurable gravity block dupe
+            if (!com.destroystokyo.paper.PaperConfig.allowGravityBlockDuplicationExploits) {
+                // Paper start - move all of this logic into portal tick
+                entity.portalWorld = ((ServerLevel) world);
+                entity.portalBlock = pos.immutable();
+                // Paper end - move all of this logic into portal tick
+            } else {
+                // CraftBukkit start - Entity in portal
+                EntityPortalEnterEvent event = new EntityPortalEnterEvent(entity.getBukkitEntity(), new org.bukkit.Location(world.getWorld(), pos.getX(), pos.getY(), pos.getZ()));
+                world.getCraftServer().getPluginManager().callEvent(event);
+
+                if (entity instanceof ServerPlayer) {
+                    ((ServerPlayer) entity).changeDimension(worldserver, PlayerTeleportEvent.TeleportCause.END_PORTAL);
+                    return;
+                }
+                // CraftBukkit end
+                entity.changeDimension(worldserver);
+            }
+            // SandPaper start - add configurable gravity block dupe
         }
 
     }
